# 2.3 Indexation

In [3]:
import pandas as pd
import numpy as np


In [4]:
data = pd.Series([0.25, 0.5, 0.75, 1.0],
                 index=['a', 'b', 'c', 'd'])

data

a    0.25
b    0.50
c    0.75
d    1.00
dtype: float64

In [5]:
data['b']

0.5

In [6]:
'a' in data


True

In [7]:
# Get the indexes
data.keys()

Index(['a', 'b', 'c', 'd'], dtype='object')

In [8]:
# The Series objects are mutables, it means we can change the value of elements

print("Before to apply a change:")
display(data)

data['a'] = 21.5
print("After made a value:")
display(data)


Before to apply a change:


a    0.25
b    0.50
c    0.75
d    1.00
dtype: float64

After made a value:


a    21.50
b     0.50
c     0.75
d     1.00
dtype: float64

In [9]:
# We can access to elements using conditionals
data [ (data > 0.3) & (data < 0.8)]

b    0.50
c    0.75
dtype: float64

In [10]:
df = pd.read_csv('data/amazonFire.csv', encoding='ISO-8859-1')

In [11]:
df.head()

year state    month  number        date
0  1998  Acre  Janeiro     0.0  1998-01-01
1  1999  Acre  Janeiro     0.0  1999-01-01
2  2000  Acre  Janeiro     0.0  2000-01-01
3  2001  Acre  Janeiro     0.0  2001-01-01
4  2002  Acre  Janeiro     0.0  2002-01-01

In [12]:
df.dtypes

year        int64
state      object
month      object
number    float64
date       object
dtype: object

Review:
- year: integer -> ok
- state: object -> strange
- month: object -> strange, why not str?
- number: float -> why not integer?
- date: object -> strange, why not datetime or str.

In [13]:
df.describe()

year       number
count  6454.000000  6454.000000
mean   2007.461729   108.293163
std       5.746654   190.812242
min    1998.000000     0.000000
25%    2002.000000     3.000000
50%    2007.000000    24.000000
75%    2012.000000   113.000000
max    2017.000000   998.000000

In [14]:
# We can access to columns in 2 different ways:
df.year

0       1998
1       1999
2       2000
3       2001
4       2002
        ... 
6449    2012
6450    2013
6451    2014
6452    2015
6453    2016
Name: year, Length: 6454, dtype: int64

In [15]:
df['year']

0       1998
1       1999
2       2000
3       2001
4       2002
        ... 
6449    2012
6450    2013
6451    2014
6452    2015
6453    2016
Name: year, Length: 6454, dtype: int64

In [16]:
# We are adding a column:
df['country'] = 'Brasil'

In [17]:
df.head()

year state    month  number        date country
0  1998  Acre  Janeiro     0.0  1998-01-01  Brasil
1  1999  Acre  Janeiro     0.0  1999-01-01  Brasil
2  2000  Acre  Janeiro     0.0  2000-01-01  Brasil
3  2001  Acre  Janeiro     0.0  2001-01-01  Brasil
4  2002  Acre  Janeiro     0.0  2002-01-01  Brasil

In [18]:
df['country'] = 'Brazil'
df.head()

year state    month  number        date country
0  1998  Acre  Janeiro     0.0  1998-01-01  Brazil
1  1999  Acre  Janeiro     0.0  1999-01-01  Brazil
2  2000  Acre  Janeiro     0.0  2000-01-01  Brazil
3  2001  Acre  Janeiro     0.0  2001-01-01  Brazil
4  2002  Acre  Janeiro     0.0  2002-01-01  Brazil

In [19]:
# To show the dimensions
df.shape

(6454, 6)

In [20]:
len(df)

6454

In [21]:
# we need to add a column with the number of months per year = 12
df['month_per_year'] = list(np.repeat(12, len(df)))
df.head()


year state    month  number        date country  month_per_year
0  1998  Acre  Janeiro     0.0  1998-01-01  Brazil              12
1  1999  Acre  Janeiro     0.0  1999-01-01  Brazil              12
2  2000  Acre  Janeiro     0.0  2000-01-01  Brazil              12
3  2001  Acre  Janeiro     0.0  2001-01-01  Brazil              12
4  2002  Acre  Janeiro     0.0  2002-01-01  Brazil              12

In [22]:
# We need to add a column with the average of fires per month:
df['fires_per_year'] = df['number'] / df['month_per_year']
df.head()


year state    month  number        date country  month_per_year  \
0  1998  Acre  Janeiro     0.0  1998-01-01  Brazil              12   
1  1999  Acre  Janeiro     0.0  1999-01-01  Brazil              12   
2  2000  Acre  Janeiro     0.0  2000-01-01  Brazil              12   
3  2001  Acre  Janeiro     0.0  2001-01-01  Brazil              12   
4  2002  Acre  Janeiro     0.0  2002-01-01  Brazil              12   

   fires_per_year  
0             0.0  
1             0.0  
2             0.0  
3             0.0  
4             0.0

In [23]:
df['fires_per_year'][0]


0.0

In [25]:
df.shape[0]

6454

In [26]:
df['fires_per_year'][0] / df.shape[0]

0.0

In [27]:
df['fires_per_month'] = df['number']>0
df.head()


year state    month  number        date country  month_per_year  \
0  1998  Acre  Janeiro     0.0  1998-01-01  Brazil              12   
1  1999  Acre  Janeiro     0.0  1999-01-01  Brazil              12   
2  2000  Acre  Janeiro     0.0  2000-01-01  Brazil              12   
3  2001  Acre  Janeiro     0.0  2001-01-01  Brazil              12   
4  2002  Acre  Janeiro     0.0  2002-01-01  Brazil              12   

   fires_per_year  fires_per_month  
0             0.0            False  
1             0.0            False  
2             0.0            False  
3             0.0            False  
4             0.0            False

In [29]:
# We can access using indexes with iloc method:
df.iloc[1,0]


1999

In [30]:
df.iloc[1, :]


year                     1999
state                    Acre
month                 Janeiro
number                    0.0
date               1999-01-01
country                Brazil
month_per_year             12
fires_per_year            0.0
fires_per_month         False
Name: 1, dtype: object

In [31]:
# If we use a range for row and column, a dataframe will return
df.iloc[:2, [1,4,-1]]


state        date  fires_per_month
0  Acre  1998-01-01            False
1  Acre  1999-01-01            False

In [32]:
# Using .loc we can get access using indexes and names for columns:
df.loc[3:4, 'date']


3    2001-01-01
4    2002-01-01
Name: date, dtype: object

In [33]:
# we can use loc method to make a filter
df.loc[ df['state'] == 'Rio' ]


year state     month  number        date country  month_per_year  \
4303  1998   Rio   Janeiro     0.0  1998-01-01  Brazil              12   
4304  1999   Rio   Janeiro     0.0  1999-01-01  Brazil              12   
4305  2000   Rio   Janeiro     0.0  2000-01-01  Brazil              12   
4306  2001   Rio   Janeiro     0.0  2001-01-01  Brazil              12   
4307  2002   Rio   Janeiro     0.0  2002-01-01  Brazil              12   
...    ...   ...       ...     ...         ...     ...             ...   
5015  2012   Rio  Dezembro    38.0  2012-01-01  Brazil              12   
5016  2013   Rio  Dezembro    62.0  2013-01-01  Brazil              12   
5017  2014   Rio  Dezembro    31.0  2014-01-01  Brazil              12   
5018  2015   Rio  Dezembro    42.0  2015-01-01  Brazil              12   
5019  2016   Rio  Dezembro    79.0  2016-01-01  Brazil              12   

      fires_per_year  fires_per_month  
4303        0.000000            False  
4304        0.000000            False  
4305        0.000000            False  
4306        0.000000            False  
4307        0.000000            False  
...              ...              ...  
5015        3.166667             True  
5016        5.166667             True  
5017        2.583333             True  
5018        3.500000             True  
5019        6.583333             True  

[717 rows x 9 columns]

In [34]:
# get the state, month, and number of fires registered in 2012
df.head()

year state    month  number        date country  month_per_year  \
0  1998  Acre  Janeiro     0.0  1998-01-01  Brazil              12   
1  1999  Acre  Janeiro     0.0  1999-01-01  Brazil              12   
2  2000  Acre  Janeiro     0.0  2000-01-01  Brazil              12   
3  2001  Acre  Janeiro     0.0  2001-01-01  Brazil              12   
4  2002  Acre  Janeiro     0.0  2002-01-01  Brazil              12   

   fires_per_year  fires_per_month  
0             0.0            False  
1             0.0            False  
2             0.0            False  
3             0.0            False  
4             0.0            False

In [35]:
df.loc[ (df['year'] == 2012 ) & (df['number'] > 0), ['state', 'month', 'number'] ]

state     month  number
54         Acre     Março    1.00
74         Acre     Abril    1.00
94         Acre      Maio    3.00
114        Acre     Junho    7.00
134        Acre     Julho   71.00
...         ...       ...     ...
6369  Tocantins    Agosto    4.28
6389  Tocantins  Setembro    7.46
6409  Tocantins   Outubro    3.68
6429  Tocantins  Novembro  179.00
6449  Tocantins  Dezembro  128.00

[313 rows x 3 columns]

In [38]:
# We can use loc and iloc methods to modify elements.
# Exm: change the mes== Dezembro by December
df.loc[ df['month'] == 'Dezembro', 'month'] = 'December'
print(df['month'].unique())

['Janeiro' 'Fevereiro' 'Março' 'Abril' 'Maio' 'Junho' 'Julho' 'Agosto'
 'Setembro' 'Outubro' 'Novembro' 'December']


In [39]:
df

year      state     month  number        date country  month_per_year  \
0     1998       Acre   Janeiro     0.0  1998-01-01  Brazil              12   
1     1999       Acre   Janeiro     0.0  1999-01-01  Brazil              12   
2     2000       Acre   Janeiro     0.0  2000-01-01  Brazil              12   
3     2001       Acre   Janeiro     0.0  2001-01-01  Brazil              12   
4     2002       Acre   Janeiro     0.0  2002-01-01  Brazil              12   
...    ...        ...       ...     ...         ...     ...             ...   
6449  2012  Tocantins  December   128.0  2012-01-01  Brazil              12   
6450  2013  Tocantins  December    85.0  2013-01-01  Brazil              12   
6451  2014  Tocantins  December   223.0  2014-01-01  Brazil              12   
6452  2015  Tocantins  December   373.0  2015-01-01  Brazil              12   
6453  2016  Tocantins  December   119.0  2016-01-01  Brazil              12   

      fires_per_year  fires_per_month  
0           0.000000            False  
1           0.000000            False  
2           0.000000            False  
3           0.000000            False  
4           0.000000            False  
...              ...              ...  
6449       10.666667             True  
6450        7.083333             True  
6451       18.583333             True  
6452       31.083333             True  
6453        9.916667             True  

[6454 rows x 9 columns]

## TEST

In [40]:
# Load the data1.csv file:
df = pd.read_csv('data/data1.csv', encoding='ISO-8859-1', parse_dates=['InvoiceDate'] )

In [41]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [42]:
df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130120 entries, 0 to 130119
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    130120 non-null  object        
 1   StockCode    130120 non-null  object        
 2   Description  129752 non-null  object        
 3   Quantity     130120 non-null  int64         
 4   InvoiceDate  130120 non-null  datetime64[ns]
 5   UnitPrice    130120 non-null  float64       
 6   CustomerID   101630 non-null  float64       
 7   Country      130120 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 7.9+ MB


Review:

- InvoiceNo              object     -> It's strange, why not integer or str
- StockCode              object     -> ok
- Description            object     -> ok
- Quantity                int64     -> ok
- InvoiceDate    datetime64[ns]     -> ok
- UnitPrice             float64     -> ok
- CustomerID            float64     -> ok
- Country                object     -> It's strange, why not str

Nulls:

- 2   Description  129752 non-null  object
- 6   CustomerID   101630 non-null  float64 


In [44]:
df.describe()

Quantity                    InvoiceDate      UnitPrice  \
count  130120.000000                         130120  130120.000000   
mean        9.658677  2011-06-09 10:48:31.827851264       4.766026   
min     -9360.000000            2010-12-01 08:26:00       0.000000   
25%         1.000000            2011-03-03 10:59:00       1.250000   
50%         4.000000            2011-07-11 11:38:00       2.100000   
75%        12.000000            2011-10-05 11:25:00       4.130000   
max      3906.000000            2011-10-20 19:51:00   13541.330000   
std        52.968237                            NaN      91.130605   

          CustomerID  
count  101630.000000  
mean    15268.647132  
min     12347.000000  
25%     13975.000000  
50%     15104.000000  
75%     16745.000000  
max     18287.000000  
std      1698.064749

In [46]:
# Which countries we have data
len(df['Country'].unique())


32

In [47]:
# How many invoices have more than 10 units
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [58]:
len(df.loc[ df['Quantity'] > 10 , 'InvoiceNo'].unique())

4089

In [59]:
# If we don't care repeat invoices 
df[df['Quantity']>10].shape[0]

33424

In [61]:
# The average over total
df[df['Quantity']>10].shape[0] / df.shape[0]


0.2568705810021519

In [65]:
# with format
p = df[df['Quantity']>10].shape[0] / df.shape[0]
print(f'The average is {p*100:.2f} %')

The average is 25.69 %


In [66]:
# Create a column with the year and filter the df with year = 2011
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [71]:
df['Year'] = df['InvoiceDate'].dt.year

In [72]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Year  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010

In [74]:
df_year = df[ df['Year'] == 2011 ]
df_year.head()

InvoiceNo StockCode                         Description  Quantity  \
17853    544292     22417      PACK OF 60 SPACEBOY CAKE CASES        24   
17854    544292     21977  PACK OF 60 PINK PAISLEY CAKE CASES        24   
17855    544292     21213         PACK OF 72 SKULL CAKE CASES        24   
17856    544292     21975      PACK OF 60 DINOSAUR CAKE CASES        24   
17857    544292     20719              WOODLAND CHARLOTTE BAG        10   

              InvoiceDate  UnitPrice  CustomerID         Country  Year  
17853 2011-02-17 12:33:00       0.55     16745.0  United Kingdom  2011  
17854 2011-02-17 12:33:00       0.55     16745.0  United Kingdom  2011  
17855 2011-02-17 12:33:00       0.55     16745.0  United Kingdom  2011  
17856 2011-02-17 12:33:00       0.55     16745.0  United Kingdom  2011  
17857 2011-02-17 12:33:00       0.85     16745.0  United Kingdom  2011

In [79]:
# How many invoices has been issued in Japan in 2010
len(df[ (df['Year'] == 2010) & (df['Country'] == 'Japan') ])

16